standalone/pruntime/enclave-api/src/lib.rs

```diff
@@ -205,11 +234,100 @@ pub mod storage_sync {
         fn validate_storage_proof(
             &self,
             state_root: Hash,
-            proof: &[u8],
+            proof: StorageProof,
             items: &[(&[u8], &[u8])],
         ) -> Result<()>;
     }
 
+    pub trait StorageSynchronizer {
+        /// Return the next block numbers to sync.
+        fn counters(&self) -> Counters;
+
+        /// Given chain headers in sequence, validate it and output the state_roots
+        fn sync_header(
+            &mut self,
+            headers: Vec<HeaderToSync>,
+            authority_set_change: Option<AuthoritySetChange>,
+        ) -> Result<chain::BlockNumber>;
+
+        /// Given the parachain headers in sequence, validate it and cached the state_roots for block validation
+        fn sync_parachain_header(
+            &mut self,
+            headers: Vec<chain::Header>,
+            proof: StorageProof,
+            storage_key: &[u8],
+        ) -> Result<chain::BlockNumber>;
+
+        /// Feed in a block of storage changes
+        fn feed_block(
+            &mut self,
+            block: &BlockHeaderWithChanges,
+            storage: &mut Storage,
+        ) -> Result<()>;
+    }
+
+    impl<Validator: BlockValidator> StorageSynchronizer for SolochainSynchronizer<Validator> {
```

Comment:
Why are there wrappers `impl` for the underlying type? Isn't it enough to just code the actual logic in `impl StorageSynchonizer for Xxx`, rather than proxying these methods to `self.<method>`?

Reply:
If we impl the logic in the `trait impl`, we need to expose the BlockSyncState, which is impl detail and less well designed, to the trait. So I think it's better to forward it.